driver-core/src/main/com/mongodb/internal/connection/OperationContext.java

```diff
@@ -15,20 +15,103 @@
  */
 package com.mongodb.internal.connection;
 
+import com.mongodb.MongoConnectionPoolClearedException;
+import com.mongodb.ServerAddress;
+import com.mongodb.connection.ClusterDescription;
+import com.mongodb.connection.ClusterType;
+import com.mongodb.connection.ServerDescription;
+import com.mongodb.lang.Nullable;
+import com.mongodb.selector.ServerSelector;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 import java.util.concurrent.atomic.AtomicLong;
+import static java.util.stream.Collectors.toList;
 
 /**
  * <p>This class is not part of the public API and may be removed or changed at any time</p>
  */
 public class OperationContext {
     private static final AtomicLong NEXT_ID = new AtomicLong(0);
     private final long id;
+    private final ServerDeprioritization serverDeprioritization;
 
     public OperationContext() {
         id = NEXT_ID.incrementAndGet();
+        serverDeprioritization = new ServerDeprioritization();
     }
 
     public long getId() {
         return id;
     }
+
+    /**
+     * @return The same {@link ServerDeprioritization} if called on the same {@link OperationContext}.
+     */
+    public ServerDeprioritization getServerDeprioritization() {
+        return serverDeprioritization;
+    }
+
+    public static final class ServerDeprioritization {
+        @Nullable
+        private ServerAddress candidate;
+        private final Set<ServerAddress> deprioritized;
+        private final DeprioritizingSelector selector;
+
+        private ServerDeprioritization() {
+            candidate = null;
+            deprioritized = new HashSet<>();
+            selector = new DeprioritizingSelector();
+        }
+
+        /**
+         * The returned {@link ServerSelector} tries to {@linkplain ServerSelector#select(ClusterDescription) select}
+         * only the {@link ServerDescription}s that do not have deprioritized {@link ServerAddress}es.
+         * If no such {@link ServerDescription} can be selected, then it selects {@link ClusterDescription#getServerDescriptions()}.
+         */
+        ServerSelector getServerSelector() {
+            return selector;
+        }
+
+        void updateCandidate(final ServerAddress serverAddress, final ClusterType clusterType) {
+            candidate = isEnabled(clusterType) ? serverAddress : null;
```

Comment:
```suggestion
            candidate = serverAddress;
```
I do not think the check is needed here. This potential "server on which the operation failed" will already be correctly excluded, because the `ServerDeprioritization` server selector is already a no-op when not sharded.

Reply:
After thinking more about this, I agree that it's fine to omit the check in the `updateCandidate` method.

Done in https://github.com/mongodb/mongo-java-driver/pull/1367/commits/246353f36dae51e28aa212488444ec54cb8ad5af.